ingest/rank_genes.py

```diff
@@ -0,0 +1,554 @@
+    # 8. Yes -- https://www.biorxiv.org/content/10.1101/2022.06.29.496888v1 -- SCP1905
+    # 9. Yes -- https://www.biorxiv.org/content/10.1101/2022.06.29.496888v1 -- SCP1903
+    # 10. Not really, but abstract mentions genes -- https://doi.org/10.1016/j.immuni.2023.01.002 -- SCP1884
+    # 11. Not really, not until 2023-10-12, but abstract mentions genes, and full text available in biorxiv -- https://doi.org/10.1093/toxsci/kfac109 -- SCP1875
+    doi_split = publication.split("https://doi.org/")
+    if len(doi_split) > 1:
+        doi = doi_split[1]
+        pmcid = fetch_pmcid(doi)
+        logger.info('Converted DOI to PMC ID: ' + pmcid)
+        text = fetch_pmcid_text(pmcid)
+    elif publication.startswith("https://www.biorxiv.org"):
+        full_text_url = f"{publication}.full.txt"
+        logger.info('full_text_url', full_text_url)
+        with urllib.request.urlopen(full_text_url) as response:
+            text = response.read().decode('utf-8')
+    else:
+        file_path = publication
+        with open(file_path) as f:
+            text = f.read()
+    return text
+
+
+def fetch_gene_cache(organism):
+    genes_filename = f"{organism}-genes.tsv"  # e.g. homo-sapiens-genes.tsv
+    base_url = "https://cdn.jsdelivr.net/npm/"
+    genes_url = f"{base_url}ideogram@1.41.0/dist/data/cache/genes/{genes_filename}.gz"
+    download_gzip(genes_url, genes_filename)
+
+    # Populate containers for various name and significance fields
+    interest_rank_by_gene = {}
+    counts_by_gene = {}
+    loci_by_gene = {}
+    full_names_by_gene = {}
+
+    i = 0
+    with open(genes_filename) as file:
+        reader = csv.reader(file, delimiter="\t")
+        for row in reader:
+            if row[0] == '#' or len(row) < 2:
+                continue
+            gene = row[4]
+            full_name = row[5]
+            counts_by_gene[gene] = 0
+            loci_by_gene[gene] = {
+                "chromosome": row[0],
+                "start": row[1],
+                "length": row[2],
+            }
+            full_names_by_gene[gene] = full_name
+            # Genes in upstream file are ordered by global popularity
+            interest_rank_by_gene[gene] = i
+            i += 1
+
+    return [interest_rank_by_gene, counts_by_gene, loci_by_gene, full_names_by_gene]
+
+def screen_false_positive_mentions(publication_text, mentions_by_gene):
+    """Screen out hits that are non-gene abbreviations
+    """
+    fp_for_tf = bool(re.search("transcription factor(s)? \(TF", publication_text))
+    if fp_for_tf:
+        del mentions_by_gene["TF"]
+        logger.info(
+            'Deleted false positive gene mention for "TF", ' +
+            'defined as "transcription factor" in publication'
+        )
+
+    # Detect gene-mentions that use have the grammatical structure:
+    #
+    #   the <term> (<"gene">)
+    #
+    # like "the Arcuate (ARC", which mentions a non-gene.
```

Comment:
I'm concerned about exclusion using heuristics. I realize that TF and ARC have many mention hits which artificially inflates their placement in ranked_genes  but the heuristics used to exclude also remove study-relevant genes. 

Gene examples:
> The glutaminergic (SLC17A6+), GABAergic (SLC32A1+), and histaminergic (HDC+) cell populations were delineated across the UMAP

1. I'm not sure the word "the" enhances selectively catching non-gene words - the gene SLC17A6 would be caught by the grammatical structure heuristic
2. I don't think any of the four instances of SLC32A1 should match the grammatical structure heuristic but SLC32A1 isn't found in ranked_genes.tsv

Exclusion by looking for "express' may also be problematic

POMC is mentioned 9 times in text (some with + and some associated with a second gene: POMC/GLP1R+) and 2 times in figure legends as well as figure text BUT four instances are in close proximity to the string "express" which excludes this gene from ranked_genes.tsv

Examples of other genes that are mentioned in the publication text but not found in ranked_genes.tsv: 

My gut feeling is that there will be a small number of strings with artificially inflated rankings while the unintended exclusions using the heuristics above could eliminate more genes whose expression should be cached. Perhaps an altermative is weighting DE results more strongly?

Reply:
That's a reasonable first interpretation.  However, it turns out those true-positive genes were excluded not based on these lexical heuristics, but rather because unrelated upstream code omitted genes not among those top-ranked by DE scores.

The "express" logic you note is actually part of a rescuing block:

```
        if is_likely_fp:
            term = fp_match.group(0)
            if (
                "express" in term or # e.g. "the Oxytocin expressing (OXT"
                gene.lower() not in term.split('(')[0].lower() # e.g. "the astrocyte (GJA1"
            ):
                continue
```

So if the candidate false-positive satisfies either of those conditions, then it is considered _not_ a false positive -- the preliminary classification as FP gets overridden.  

The only gene-like tokens in https://doi.org/10.1126/sciadv.adf6251 excluded by `screen_false_positive_mentions` are "TF" and "ARC".  You can verify this by looking in your analog of the example below:

```
(env) wmce2-f16:scp-ingest-pipeline eweitz$ grep "false positive" rank_genes_2024-01-03T162344_log.txt
2024-01-03T16:24:03-0500 rank_genes INFO:Deleted false positive gene mention for "TF", defined as "transcription factor" in publication
2024-01-03T16:24:03-0500 rank_genes INFO:Deleted false positive gene mention for "ARC", defined as "the Arcuate (ARC" in publication
```

So the heuristics seem accurate in this case.  That said, your particular finding that `rank_genes.py` didn't include some frequently-mentioned genes was insightful detective work, and key to fixing the bug's upstream cause!